### DISCUSSION TITLE
Is it safe to use SerialPIO with FreeRTOS SMP?

### DISCUSSION THREAD

**Discussion by agent106** [earlephilhower/arduino-pico#1857] [General]
The FreeRTOS doc page [](https://arduino-pico.readthedocs.io/en/latest/freertos.html) says that 
"While the core now supports FreeRTOS, most (probably all) Arduino libraries were not written to support preemptive multithreading."

How about SerialPIO? I want to create two separate tasks that use a few software serial UART interfaces and if I use mutexes to guard each individual SerialPIO instance - should it be okay? In other words, is it safe to use different software UARTS simultaneously from different threads.

Thank you!!

**Comment by earlephilhower** [comment]
It should be safe, and if it's not then I'm sure it can be made so.  (It's not tested, though, so be prepared to debug a little!)

If you're only using a single thread to talk to each SerialPIO there shouldn't be a need to use a mutex to control access.  In fact, there's already one (like in the SerialUSB/UART) to enable both cores to write to it.

The only thing I can see that could be a problem is calling `SerialPIO::begin()` in multiple threads.  The issue is there's shared infrastructure below them (`PIOProgram` and some static arrays) which are not protected and could get scrambled if run on both cores simultaneously or partially started in 1 task and then restarted in a 2nd swapped in task.  So, if you can `SerialPIO::begin()` from a single task and not the individual tasks I think that would be safe.

**Reply by agent106** [reply]
Got it, thank you! I will try that, but meanwhile, I'm running into another issue. I want to use 4 SerialPIO connections and I need a separate WiFi connection. 4 soft serial connections seem to work wine, but as soon as I add WiFi connection - the Pico freezes. However, this only happens when I initialize WiFI **after** setting up the soft serial - if WiFi is initialized **first** and then I init the 4 serials - everything is okay. Another weirdness is that the problem goes away if I only have 3 soft serials.

Here is the test code that reproduces this:

```
#include <SoftwareSerial.h>
#include <WiFi.h>

const char *wifi_ssid = "*****";
const char *wifi_pass = "*****";

const int fifo_size = 256;
SerialPIO soft_serial1(15, 14, fifo_size);
SerialPIO soft_serial2(13, 12, fifo_size);
SerialPIO soft_serial3(11, 10, fifo_size);
SerialPIO soft_serial4(7, 6, fifo_size);

void initWifi() {
  Serial.print("Connecting WiFi...");
  WiFi.begin(wifi_ssid, wifi_pass);
  while (WiFi.status() != WL_CONNECTED) {
    delay(500);
    Serial.print(".");
  }
  Serial.print("OK IP: ");
  Serial.println(WiFi.localIP());    
}

void setup() {
  Serial.begin(115200);  
  Serial.print("Waiting to start");
  for(int i=0; i<10; i++) {
    delay(1000);
    Serial.print(".");    
  }
  Serial.println("Starting!");


  // The problem
  soft_serial1.begin(115200, SERIAL_8N1);
  soft_serial2.begin(115200, SERIAL_8N1);
  soft_serial3.begin(115200, SERIAL_8N1);
  soft_serial4.begin(115200, SERIAL_8N1); // COMMENT THIS OUT - AND IT WORKS!!!
  
  initWifi();

  // ---------------------------------
}

void loop() {
  Serial.println("I'm okay!");
  delay(1000);
}
```
To recap, what fails is: init 4 serials before WiFi. To fix, either init the WiFi before serials, or reduce number of serials to 3 or less. The reason I need to init WiFi after I init serials, is that I don't always need WiFI connection - I want to be able to connect and disconnect in loop().

Any ideas? Thanks again!

**Reply by earlephilhower** [reply]
It's just not possible to use WiFi and 4 soft serials.  You're out of PIO state machines.  Each serial port needs 2 SMs, and the WiFi core from RPi also needs 2 (I believe...in any case at least one).  There are 8 total on-chip, 4 per PIO.

The crash is really a `panic` in the RPI core when it's not able to allocate a PIO SM to run the soft-SPI that connects up the WiFi chip.  If you hook up a Picoprobe you should see the call.

Not really anything you can do about it other than, if possible, only use 1-directional soft-serial port (i.e. if you're only ever sending or only ever receiving) which only needs 1 SM.

**Reply by agent106** [reply]
Ah! Sorry, I didn't use Picoprobe - my bad! It makes perfect sense - thank you very much!

**Reply by agent106** [reply]
@earlephilhower Do hardware serials use state machines? I mean, can I have 2 hardware UARTs, 3 software and a WiFi connection?

**Reply by earlephilhower** [reply]
Serial1 and Serial2 use the hardware UARTs, no PIO SM required.  So you can have 3 soft serials, 2 hard serials, and wifi if you're so inclined, like you suggest.
